---
title: "human-breast-cancer-analysis"
author: "Melody"
date: "`r Sys.Date()`"
output: workflowr::wflow_html
editor_options:
  chunk_output_type: console
---
# Introduction
The dataset used in this vignette is the Xenium human breast cancer.
The data can be accessed here:
<https://www.10xgenomics.com/products/xenium-in-situ/preview-dataset-human-breast>


Details about each file can be found here:
<https://www.10xgenomics.com/support/in-situ-gene-expression/documentation/steps/onboard-analysis/understanding-xenium-outputs#overview>


```{r global options, include = FALSE}
options(width = 80, digits = 3) # output length
knitr::opts_chunk$set(cache = FALSE,echo = TRUE,prompt = FALSE, 
                      comment = NA,message = FALSE,warning = FALSE,
                      fig.align = "center", fig.keep= "all",fig.show ="hold")

# wx=7
#wy=5
```

```{r}
library(jazzPanda)
library(SpatialExperiment)
library(Seurat)
library(ggplot2)
library(data.table)
library(spatstat)
library(dplyr)
library(glmnet)
library(caret)
library(ComplexUpset)
library(corrplot)
library(igraph)
library(ggraph)
library(ggrepel)
library(gridExtra)
library(patchwork)
library(RColorBrewer)
library(ggvenn)
#library(multtest)
#library(metap)
library(scCustomize)
library(tidyr)
library(xtable)
```

```{r}
# calculate peak memory
PEAK_MEM_ID=which(colnames(gc())=="max used")
PEAK_MEM_MB = PEAK_MEM_ID+1

PA = "figures/main/figure3_xenium_hbreast/"
# theme setting for figures
defined_theme <- theme(strip.text = element_text(size = rel(1)),
                      axis.line=element_blank(),
                      axis.ticks=element_blank(),
                      axis.text=element_blank(),
                      legend.position = "none",
                      axis.title=element_blank(),
                      panel.background=element_blank(),
                      panel.grid.major=element_blank(),
                      panel.grid.minor=element_blank(),
                      plot.background=element_blank())

# A help function to load Xenium data
get_xenium_data<-function(path,mtx_name, trans_name="transcript_info.csv.gz",
                    cells_name="cell_info.csv.gz"){

    transcript_info <- as.data.frame(fread(paste(path, trans_name,sep="")))
    cell_info <- as.data.frame(fread(paste(path,cells_name,sep="")))

    data <- Read10X(data.dir = paste(path,mtx_name, sep=""))

    cm <- as.matrix(data$`Gene Expression`)
    r_codeword <- as.matrix(data$`Negative Control Codeword`)

    r_probe <- as.matrix(data$`Negative Control Probe`)
    # merge negative control genes and real genes
    cm_neg <- as.data.frame(rbind(r_probe, r_codeword))
    zero_cells <- colnames(cm)[colSums(cm)==0]

    transcript_info$x <- as.numeric(transcript_info$x_location)
    transcript_info$y <- as.numeric(transcript_info$y_location)

    cell_info$x <- as.numeric(cell_info$x_centroid)
    cell_info$y <- as.numeric(cell_info$y_centroid)

    return (list(cm = cm, cm_neg=cm_neg, zero_cells = zero_cells,
                    trans_info=transcript_info, cell_info=cell_info,
                    probe = row.names(r_probe),
                    codeword=row.names(r_codeword)))

}
```

# Load sample1 and sample2 dataset
```{r load rep1 data and clustering, fig.width=8, fig.height=6}

sp1_path <- "/stornext/Bioinf/data/lab_phipson/data/xenium_human_breast_cancer/rep1/"

sp1 = get_xenium_data(sp1_path, 
                mtx_name="cell_feature_matrix",
                trans_name="transcripts.csv.gz", 
                cells_name="cells.csv.gz" )
sp1$trans_info = sp1$trans_info[sp1$trans_info$qv >=20 & 
                            sp1$trans_info$cell_id != -1 & 
                            !(sp1$trans_info$cell_id %in% sp1$zero_cells), ]

## sample2
sp2_path <- "/stornext/Bioinf/data/lab_phipson/data/xenium_human_breast_cancer/sample2/"
sp2 = get_xenium_data(sp2_path, 
                mtx_name="cell_feature_matrix",
                trans_name="transcripts.csv.gz", 
                cells_name="cells.csv.gz" )

sp2$trans_info = sp2$trans_info[sp2$trans_info$qv >=20 & 
                        sp2$trans_info$cell_id != -1 & 
                        !(sp2$trans_info$cell_id %in% sp1$zero_cells), ]


gene_info = read.table(file ="data/Xenium_FFPE_Human_Breast_Cancer_Rep1_panel.tsv",
                       sep = '\t', header = TRUE)

gene_info$label = paste(gene_info$Name, gene_info$Annotation, sep=" - ")
gene_info$Annotation = factor(gene_info$Annotation)
sum_df = gene_info[,c("Name", "Annotation"),drop=FALSE] %>% 
  group_by(Annotation) %>% dplyr::count() %>% data.frame()

ggplot(sum_df, aes(x = Annotation, y = n, fill=n)) +
    geom_bar(stat = "identity", position="stack", fill="black") +
    geom_text(aes(label = n),size=3, 
              position = position_dodge(width = 0.9), vjust = -0.5) + 
    scale_y_continuous(expand = c(0,0), limits = c(0, 60))+
    theme_bw()+
    defined_theme+theme(axis.text.x=element_text(angle=90, vjust = 0.4,hjust=1))+
    labs(title = " ", x = " ", y = "number of genes", fill="Annotation") 
```

```{r}
#### check memory usage 
gc(reset = TRUE)
load_cm_imem <- sum(gc()[, 6])
data_sp1 <- Read10X(data.dir=paste(sp1_path, "cell_feature_matrix/", sep=""))
cm1 <- as.matrix(data_sp1$`Gene Expression`)

data_sp2 <- Read10X(data.dir=paste(sp2_path, "cell_feature_matrix/", sep=""))
cm2 <- as.matrix(data_sp2$`Gene Expression`)    
load_cm_fmem <- sum(gc()[, PEAK_MEM_MB])
load_cm_dmem <- load_cm_fmem - load_cm_imem

cat("loading count matrix data for",nrow(sp1$cm),"genes and",
    ncol(cm1)+ncol(cm2),"cells took",round(load_cm_dmem / (1024),5),"GB memory")
rm(data_sp1, data_sp2, cm1,cm2)
## check transcripts memory
gc(reset = TRUE)
load_tr_imem <- sum(gc()[, PEAK_MEM_MB])

transcript_info1<- as.data.frame(fread(paste(sp1_path,"transcripts.csv.gz", sep="")))
transcript_info2<- as.data.frame(fread(paste(sp2_path,"transcripts.csv.gz", sep="")))
load_tr_fmem <- sum(gc()[, PEAK_MEM_MB])
load_tr_dmem <- load_tr_fmem - load_tr_imem
rm(transcript_info1,transcript_info2)

cat("loading transcript coordinates data for",nrow(sp1$cm),"genes and",
    ncol(sp1$cm)+ncol(sp2$cm),"cells took",
    round(load_tr_dmem / (1024),5),"GB memory")

## check cell metadata memory 
gc(reset = TRUE)
load_cell_imem <- sum(gc()[, PEAK_MEM_MB])
cells_meta1 =  as.data.frame(fread(paste(sp1_path,"cells.csv.gz", sep=""))) 
cells_meta2 =  as.data.frame(fread(paste(sp2_path,"cells.csv.gz", sep=""))) 
load_cell_fmem <- sum(gc()[, PEAK_MEM_MB])
load_cell_dmem <- load_cell_fmem - load_cell_imem
cat("loading cell coordinates data for",
    ncol(sp1$cm)+ncol(sp2$cm),"cells took",round(load_cell_dmem / (1024),5),"GB memory")
rm(cells_meta1,cells_meta2)

```


# Negative controls
## total detections of each negative control
```{r falsecode, fig.width=10, fig.height=10}

probe_coords <- as.data.frame(rbind(cbind(sample="sample1",sp1$trans_info[sp1$trans_info$feature_name %in% sp1$probe, c("feature_name","x","y")]),
                                        cbind(sample="sample2",sp2$trans_info[sp2$trans_info$feature_name %in% sp2$probe, c("feature_name","x","y")])
                                        ))



codeword_coords = as.data.frame(rbind(cbind(sample="sample1",sp1$trans_info[sp1$trans_info$feature_name %in% sp1$codeword, c("feature_name","x","y")]),
                                        cbind(sample="sample2",sp2$trans_info[sp2$trans_info$feature_name %in% sp2$codeword, c("feature_name","x","y")])
                                        ))
```



# Clustering 
```{r fig.width= 8, fig.height= 4}
graphclust_sp1=read.csv("data/Xenium_rep1_supervised_celltype.csv")
graphclust_sp1$anno =as.character(graphclust_sp1$Cluster)
target_clusters = c("Tumor", "Stromal","Macrophages","Myoepithelial",
                    "T_Cells", "B_Cells","Endothelial", "Dendritic", "Mast_Cells")

t_cells =  c("CD4+_T_Cells","CD8+_T_Cells","T_Cell_&_Tumor_Hybrid",
             "Stromal_&_T_Cell_Hybrid")
dc_cells = c("LAMP3+_DCs","IRF7+_DCs")
macro_cells = c("Macrophages_1","Macrophages_2")
myo_cells = c("Myoepi_KRT15+", "Myoepi_ACTA2+")
tumor_cells = c("Invasive_Tumor", "Prolif_Invasive_Tumor")
graphclust_sp1[graphclust_sp1$Cluster %in% t_cells, "anno"] = "T_Cells"
graphclust_sp1[graphclust_sp1$Cluster %in% macro_cells,"anno"] = "Macrophages"
graphclust_sp1[graphclust_sp1$Cluster %in% c("DCIS_1", "DCIS_2"),"anno"] = "DCIS"
graphclust_sp1[graphclust_sp1$Cluster %in% dc_cells,"anno"] = "Dendritic"
graphclust_sp1[graphclust_sp1$Cluster %in% myo_cells,"anno"] = "Myoepithelial"
graphclust_sp1[graphclust_sp1$Cluster %in% tumor_cells,"anno"] = "Tumor"

graphclust_sp1$anno=factor(graphclust_sp1$anno,
                                  levels=c("Tumor", "DCIS", "Stromal",
                                           "Macrophages","Myoepithelial",
                                           "T_Cells", "B_Cells",
                                           "Endothelial", 
                                           "Dendritic", "Mast_Cells",
                                           "Perivascular-Like","Unlabeled"))


sp1_clusters = as.data.frame(cbind(as.character(graphclust_sp1$anno),
                              paste("c",as.numeric(factor(graphclust_sp1$anno)), 
                                    sep="")))

row.names(sp1_clusters) = graphclust_sp1$Barcode
colnames(sp1_clusters) = c("anno","cluster")

cells= sp1$cell_info
row.names(cells) =cells$cell_id
rp_names =  row.names(sp1_clusters) 
sp1_clusters[rp_names,"x"] = cells[match(rp_names,row.names(cells)),
                                             "x"]
sp1_clusters[rp_names,"y"] = cells[match(rp_names,row.names(cells)),
                                             "y"]

sp1_clusters$anno=factor(sp1_clusters$anno,
                          levels=c("Tumor", "DCIS", "Stromal",
                                   "Macrophages","Myoepithelial",
                                   "T_Cells", "B_Cells",
                                   "Endothelial", 
                                   "Dendritic", "Mast_Cells",
                                   "Perivascular-Like","Unlabeled"))
sp1_clusters$cluster=factor(sp1_clusters$cluster,
                             levels=paste("c", 1:12, sep=""))

sp1_clusters$cells =paste(row.names(sp1_clusters),"_1",sep="")
sp1_clusters$sample="sample1"
selected_sp1 = sp1_clusters
selected_sp1$cluster_comb = as.character(selected_sp1$anno)
selected_sp1[selected_sp1$cluster_comb =="DCIS","cluster_comb"]  = "Tumor"
selected_sp1 = selected_sp1[selected_sp1$cluster_comb %in% target_clusters,]
selected_sp1$cluster_comb = factor(selected_sp1$cluster_comb, levels=target_clusters)
selected_sp1$cluster =paste("c",as.numeric(factor(selected_sp1$cluster_comb)),
                                     sep="")
selected_sp1 = selected_sp1[,c("cluster","x","y","cells","sample","cluster_comb")]
colnames(selected_sp1)[6]="anno"

sample2_anno = read.csv("data/Sample2_Xenium_cell_type_manual.csv", row.names = 1)
sp2_clusters = as.data.frame(cbind(row.names(sample2_anno),
                                   as.character(sample2_anno$cell_type)))
colnames(sp2_clusters) = c("cells","anno")
row.names(sp2_clusters) = sp2_clusters$cells
cells= sp2$cell_info
row.names(cells) = cells$cell_id
rp_names = sp2_clusters$cells
sp2_clusters[rp_names,"x"] = cells[match(rp_names,row.names(cells)),
                                             "x"]
sp2_clusters[rp_names,"y"] = cells[match(rp_names,row.names(cells)),
                                             "y"]

sp2_clusters$sample="sample2"

sp2_clusters[sp2_clusters$anno %in% c("B", "Plasma"),"anno"]="B_Cells" 
sp2_clusters[sp2_clusters$anno %in% c("Macrophage"),"anno"]="Macrophages" 
sp2_clusters[sp2_clusters$anno %in% c("T","NK"),"anno"]="T_Cells" 
sp2_clusters[sp2_clusters$anno %in% c("Fibroblast"),"anno"]="Stromal" 
sp2_clusters[sp2_clusters$anno %in% c("Mast"),"anno"]="Mast_Cells" 
sp2_clusters$anno = factor(sp2_clusters$anno,
                           levels=c("Tumor", "Stromal","Macrophages","Myoepithelial",
                    "T_Cells", "B_Cells","Endothelial", "Dendritic", "Mast_Cells","ML","LP","Adipocyte"))

sp2_clusters$cells=paste(sp2_clusters$cells,"-sp2", sep="")


sp2_clusters$anno = as.character(sp2_clusters$anno)
selected_sp2 = sp2_clusters[sp2_clusters$anno %in% target_clusters,]

selected_sp2$anno = factor(selected_sp2$anno, levels=target_clusters)
selected_sp2$cluster =paste("c",as.numeric(factor(selected_sp2$anno)), 
                                    sep="")

selected_sp2 = selected_sp2[,c("cluster","x","y","cells","sample","anno")]
clusters = rbind(selected_sp1, selected_sp2)
table(clusters$sample, clusters$anno)
clusters$anno = factor(clusters$anno, target_clusters)


ct_names =c("Tumor", "Stromal","Macrophages","Myoepithelial", "T_Cells", 
            "B_Cells","Endothelial", "Dendritic", "Mast_Cells")
clusters$anno = factor(clusters$anno, levels = ct_names)


my_colors<- c("#FC8D62","#66C2A5" ,"#8DA0CB","#E78AC3",
              "#A6D854","skyblue","purple3","#E5C498","blue")


```

# Figure 3(b)
```{r}
sum_df = clusters[,c("anno", "cells","sample"),drop=FALSE] %>% group_by(anno, sample) %>% dplyr::count()

pdf(paste(PA, "figure3b_xenium_hbreast_celltype_prop.pdf", sep=""), width=6, height=5)
ggplot(sum_df, aes(x = sample, y = n, fill=anno)) +
  geom_bar( stat="identity",position="fill") +
  scale_y_continuous(labels = scales::percent,expand=c(0,0))+
  labs(title = " ", x = " ", y = "proportion of cell types", fill="cell types") +
     scale_fill_manual(values = c("#FC8D62","#66C2A5" ,"#8DA0CB","#E78AC3",
                               "#A6D854","skyblue","purple3","#E5C498","blue"))+
   theme(legend.position = "right",axis.text.x= element_text(size=15),
         strip.text = element_text(size = rel(1)),
          axis.line=element_blank(),
               axis.text.y=element_blank(),axis.ticks=element_blank(),
               axis.title.x=element_blank(),
               axis.title.y=element_blank(),
               panel.background=element_blank(),
               panel.grid.major=element_blank(),
               panel.grid.minor=element_blank(),
         plot.background=element_blank())
dev.off()
```
# Figure 3(a)
```{r fig.width= 10, fig.height= 3}

p1<- ggplot(data = clusters,
        aes(x = x, y = y, color=anno))+
        geom_point(size=0.001, alpha=0.7)+
        facet_wrap(~sample, nrow=1)+
        theme_classic()+
        scale_y_reverse()+
        scale_color_manual(values = c("#FC8D62","#66C2A5" ,"#8DA0CB","#E78AC3",
                               "#A6D854","skyblue","purple3","#E5C498","blue"))+
        guides(color=guide_legend(title="cluster", ncol = 1,
        override.aes=list(alpha=1, size=8)))+ defined_theme+
  theme(strip.background = element_rect(color="white"),
        legend.position = "none",legend.key.width = unit(1, "cm"),
                    legend.title = element_text(size = 12),
                    legend.text = element_text(size = 8),
                    legend.key.height = unit(1, "cm"),
        strip.text = element_text(size = 12))

jpeg(paste(PA, "figure3a_xenium_hbreast_clustering_xy.jpg", sep=""), 
     width=2000, height=900, res=200)
p1
dev.off()

```

When we have multiple samples in the dataset, we can find shared marker genes 
by providing sample information in the cluster input for \code{lasso_markers}.

# Find shared marker genes for sample1 and sample2
## jazzPanda
```{r}

shared_genes = intersect(row.names(sp1$cm), row.names(sp2$cm))
seed_number= 589

grid_length=40
gc(reset = TRUE)
sv_st = Sys.time()
sv_imem <- sum(gc()[, PEAK_MEM_MB])
keep_targets = c(shared_genes,
                 intersect(sp1$probe,sp2$probe),
                 intersect(sp1$codeword,sp2$codeword))




# get spatial vectors
sp1_sp2_vectors = get_vectors(x= list("sample1" = sp1$trans_info, 
                                      "sample2" = sp2$trans_info),
                              sample_names = c("sample1", "sample2"),
                           cluster_info = clusters, bin_type="square",
                           bin_param=c(grid_length,grid_length), 
                           test_genes = shared_genes,  
                           n_cores = 5)
sv_ed= Sys.time()


sv_fmem <-sum(gc()[, PEAK_MEM_MB])
sv_dmem <- sv_fmem - sv_imem
sv_time = difftime(sv_ed,sv_st, units = "mins")
cat("creating vectors for",nrow(sp1$cm),"genes and",
    length(unique(clusters$cluster)),"clusters took", sv_time,"mins and",
    round(sv_dmem / (1024),5),"GB memory")

# create noise vectors 

sp1_sp2_nc_vectors = create_genesets(x= list("sample1" = sp1$trans_info, 
                                      "sample2" = sp2$trans_info),
                            sample_names = c("sample1", "sample2"),
                            name_lst=list(probe=intersect(sp1$probe,sp2$probe), 
                                codeword=intersect(sp1$codeword,sp2$codeword)),
                            bin_type="square",
                            bin_param=c(40,40), 
                            cluster_info = NULL)

set.seed(seed_number)
gc(reset=TRUE)
glm_st = Sys.time()
glm_imem <-sum(gc()[, PEAK_MEM_MB])
sp1_sp2_lasso_with_nc = lasso_markers(gene_mt=sp1_sp2_vectors$gene_mt,
                                cluster_mt = sp1_sp2_vectors$cluster_mt,
                                sample_names=c("sample1","sample2"),
                                keep_positive=TRUE, 
                                background=sp1_sp2_nc_vectors)
glm_fmem <- sum(gc()[, PEAK_MEM_MB])
glm_ed <- Sys.time()
glm_dmem <- glm_fmem - glm_imem
glm_time <- difftime(glm_ed, glm_st, units = "mins")
cat("jazzPanda-linear model took", glm_time,"mins and",
    round(glm_dmem / (1024),5),"GB memory")

sp1_sp2_res =get_top_mg(sp1_sp2_lasso_with_nc, coef_cutoff=0.1)
sp1_sp2_res$celltype ="UNK"
clusters$anno = as.character(clusters$anno)
sp1_sp2_res$celltype = clusters[match(sp1_sp2_res$top_cluster,clusters$cluster),"anno"]
sp1_sp2_res[is.na(sp1_sp2_res$celltype),"celltype"] = "UNK"
sp1_sp2_res$Annotation = gene_info[match(sp1_sp2_res$gene,gene_info$Name),
                            "Annotation"]
table(sp1_sp2_res$top_cluster)
cluster_names = unique(as.character(sp1_sp2_res$celltype))
sp1_sp2_res$celltype = as.character(sp1_sp2_res$celltype)

jazzPanda_summary = as.data.frame(table(sp1_sp2_res$top_cluster))
colnames(jazzPanda_summary) = c("cluster","jazzPanda_glm")


```



# Figure 2(d) 
## spatial coordinates for one cluster
```{r}
cl = "T_Cells"
p_cl<- ggplot(data = clusters[clusters$anno==cl, ],
        aes(x = x, y = y))+ 
        facet_wrap(~sample, nrow=2)+
        #geom_hex(bins = 120)+
        geom_point(size=0.01, alpha=0.7,color="#A6D854")+
        guides(fill = guide_colorbar(height= unit(0.6, "cm")))+
        scale_fill_gradient(low="white", high="black") + 
        #scale_fill_gradient(low="white", high="maroon4") + 
        defined_theme+ 
        theme(legend.position = "bottom",
              strip.background = element_rect(colour = "white", 
                                              fill = "white", 
                                              linetype = "solid"), 
              panel.border = element_rect(colour = "NA", fill="NA"),
               strip.text = element_text(size = 15),
              panel.spacing = unit(0, "lines"))


jpeg(paste(PA, "figure3c_xenium_hbreast_tcell_xy.jpg", sep=""),
      width = 1000, height = 1800, res=200)
p_cl 
dev.off()
```
# Figure 2(e) 
## spatial coordinates for top marker gene
```{r top3-gene-vis-xy-perm, fig.width=10, fig.height=12}
cl = "T_Cells"
inters = "IL7R"

iters_sp1= sp1$trans_info$feature_name %in% inters
vis_r1 =sp1$trans_info[iters_sp1,
                        c("x","y","feature_name")]
vis_r1$sample="sample1"
iters_rep2= sp2$trans_info$feature_name %in% inters
vis_r2 =sp2$trans_info[iters_rep2,
                        c("x","y","feature_name")]
vis_r2$sample="sample2"
p1<- ggplot(data = vis_r1,
            aes(x = x, y = y))+ 
            #geom_hex(bins = tr_hex[cl_ids])+
            geom_point(size=0.01, alpha=0.7)+
            facet_wrap(~sample, scales="free", ncol=1)+
            #scale_fill_gradient(low="white", high="maroon4") + 
            guides(fill = guide_colorbar(height= unit(0.6, "cm")))+
            defined_theme+  theme(legend.position = "bottom",
              strip.background = element_rect(colour = "white", 
                                              fill = "white", 
                                              linetype = "solid"), 
              panel.border = element_rect(colour = "NA", fill="NA"),
              strip.text = element_text(size = 15),
              panel.spacing = unit(0, "lines"))

p2<- ggplot(data = vis_r2,
            aes(x = x, y = y))+ 
            #geom_hex(bins = tr_hex[cl_ids])+
            geom_point(size=0.01, alpha=0.7)+
            facet_wrap(~sample, scales="free", ncol=1)+
            #scale_fill_gradient(low="white", high="maroon4") + 
            guides(fill = guide_colorbar(height= unit(0.6, "cm")))+
            defined_theme+  theme(legend.position = "bottom",
              strip.background = element_rect(colour = "white", 
                                              fill = "white", 
                                              linetype = "solid"), 
              panel.border = element_rect(colour = "NA", fill="NA"),
              strip.text = element_text(size = 15),
              panel.spacing = unit(0, "lines"))

lyt = (p1 / p2) 
layout_design <- lyt + patchwork::plot_layout(heights = c(1,1),widths = c(1, 1)) 
jpeg(paste(PA, "figure3d_xenium_hbreast_top1_xy_glm.jpg", sep=""),
     width = 1000, height = 1800, res=200)
layout_design
dev.off()
```
# Figure 2(f) 

```{r}

cl = "c5"
mk_gene ="IL7R"

dff = as.data.frame(cbind(sp1_sp2_vectors$cluster_mt[,cl],sp1_sp2_vectors$gene_mt[,mk_gene]))
colnames(dff) = c("cluster", mk_gene)
dff$sample= "sample1"
dff[1600:3200,"sample"] = "sample2"
dff$vector_id = c(1:1600, 1:1600)
long_df <- dff %>% 
pivot_longer(cols = -c(cluster, sample, vector_id), names_to = "gene", 
           values_to = "vector_count")
long_df$gene = factor(long_df$gene, levels=mk_gene)
dff$sample = factor(dff$sample , levels=c("sample1","sample2"))
p<-ggplot(long_df, aes(x = cluster, y = vector_count, color =gene )) +
geom_point(color="black", size=0.01) +
facet_wrap(~sample, nrow=2) +
labs(x = paste("T cell", " cluster vector ", sep=""), y = "IL7R gene vector") +
theme_minimal()+
    scale_y_continuous(expand = c(0.01,0.01))+ 
    scale_x_continuous(expand =  c(0.01,0.01))+ 
theme(panel.grid = element_blank(),legend.position = "none",
    strip.text = element_blank(),#element_text(size = rel(1)),
    axis.line=element_blank(),
    axis.text=element_text(size = 12),
    axis.ticks=element_line(color="black"),
    axis.title=element_text(size = 15),
    panel.border  =element_rect(colour = "black", fill=NA, linewidth=0.5)
    )

pdf(paste(PA, "figure3e_xenium_hbreast_vvplot.pdf", sep=""),
      height = 9, width = 5)
p
dev.off()
```


# limma
```{r}
limma_st = Sys.time()
gc(reset = TRUE)
limma_imem <-sum(gc()[, PEAK_MEM_MB])

cm_new1=sp1$cm[,setdiff(colnames(sp1$cm),c(sp1$zero_cells,"cate"))]
colnames(cm_new1)=paste(colnames(cm_new1),"_1", sep="")
biorep1=factor(rep(c("hb1"),c(ncol(cm_new1))))
names(biorep1) = colnames(cm_new1)

cm_sp2=sp2$cm[,setdiff(colnames(sp2$cm),c(sp2$zero_cells,"cate"))]
colnames(cm_sp2)=paste(colnames(cm_sp2),"-sp2", sep="")
biosp2=factor(rep(c("hb2"),c(ncol(cm_sp2))))
names(biosp2) = colnames(cm_sp2)

rep_clusters = clusters
y=DGEList(cbind(cm_new1[shared_genes,intersect(colnames(cm_new1),rep_clusters$cells)], cm_sp2[shared_genes,intersect(colnames(cm_sp2),rep_clusters$cells)]))
sample=c(biorep1[intersect(colnames(cm_new1),rep_clusters$cells)],biosp2[intersect(colnames(cm_sp2),rep_clusters$cells)])
logcounts.all <- normCounts(y,log=TRUE,prior.count=0.1)
all.ct <- factor(rep_clusters$cluster)

design <- model.matrix(~0+all.ct+sample)
colnames(design)[1:(length(levels(all.ct)))] <- levels(all.ct)

mycont <- matrix(0,ncol=length(levels(all.ct)),nrow=length(levels(all.ct)))
colnames(mycont)<-levels(all.ct)
diag(mycont)<-1
mycont[upper.tri(mycont)]<- -1/(length(levels(all.ct))-1)
mycont[lower.tri(mycont)]<- -1/(length(levels(all.ct))-1)

# Fill out remaining rows with 0s
zero.rows <- matrix(0,ncol=length(levels(all.ct)),nrow=(ncol(design)-length(levels(all.ct))))
test <- rbind(mycont,zero.rows)

fit <- lmFit(logcounts.all,design)
fit.cont <- contrasts.fit(fit,contrasts=test)
fit.cont <- eBayes(fit.cont,trend=TRUE,robust=TRUE)

fit.cont$genes <- shared_genes

#treat_res <- treat(fit.cont,lfc=0.5)
limma_dt<-decideTests(fit.cont)
#limma_dt
summary(limma_dt)

limma_fmem <-sum(gc()[, PEAK_MEM_MB])
limma_dmem <- limma_fmem - limma_imem
limma_ed = Sys.time()
limma_time=difftime(limma_ed, limma_st, units = "mins")
cat("limma took",limma_time ,"mins and",
    round(limma_dmem / (1024),5),"GB memory")

limma_summary = as.data.frame(summary(limma_dt))
limma_summary = limma_summary[limma_summary$Var1=="Up",c(2,3)]
colnames(limma_summary) = c("cluster","limma")
```

# FindMarkers 
```{r}

set.seed(1939)
cm_new1=sp1$cm[,setdiff(colnames(sp1$cm),c(sp1$zero_cells,"cate"))]
colnames(cm_new1)=paste(colnames(cm_new1),"_1", sep="")
biorep1=factor(rep(c("hb1"),c(ncol(cm_new1))))
names(biorep1) = colnames(cm_new1)

sp1_seu=CreateSeuratObject(counts = cm_new1, project = "hb1")
sp1_seu=AddMetaData(object=sp1_seu, metadata = biorep1, col.name="biorep")
DefaultAssay(sp1_seu) <- 'RNA'
sp1_seu = NormalizeData(sp1_seu,normalization.method ="LogNormalize")
sp1_seu = FindVariableFeatures(sp1_seu, selection.method = "vst",
                                nfeatures = nrow(sp1$cm), verbose = FALSE)

sp1_seu=ScaleData(sp1_seu)

sp1_seu=RunPCA(sp1_seu, npcs = 50, verbose = FALSE)

set.seed(1939)
cm_sp2=sp2$cm[,setdiff(colnames(sp2$cm),c(sp2$zero_cells,"cate"))]
colnames(cm_sp2)=paste(colnames(cm_sp2),"_2", sep="")
biosp2=factor(rep(c("hb2"),c(ncol(cm_sp2))))
names(biosp2) = colnames(cm_sp2)
sp2_seu=CreateSeuratObject(counts = cm_sp2, project = "hb2")
sp2_seu=AddMetaData(object=sp2_seu, metadata = biosp2, col.name="biorep")
DefaultAssay(sp2_seu) <- 'RNA'
sp2_seu <- NormalizeData(sp2_seu,normalization.method ="LogNormalize")
sp2_seu = FindVariableFeatures(sp2_seu, selection.method = "vst",
                                nfeatures = nrow(sp2$cm), verbose = FALSE)

sp2_seu <- ScaleData(sp2_seu)
sp2_seu=RunPCA(sp2_seu, npcs = 50, verbose = FALSE)

fm_st = Sys.time()
gc(reset = TRUE)
fm_imem <-sum(gc()[, PEAK_MEM_MB])

merged_seu = Merge_Seurat_List(
  list(sp1_seu[shared_genes,], sp2_seu[shared_genes,]),
 # add.cell.ids = c(),
  merge.data = FALSE,
  project = "hbreast"
)
Idents(merged_seu) = clusters$cluster


# re-join layers after integration
merged_seu[["RNA"]] <- JoinLayers(merged_seu[["RNA"]])

FM_res <- FindAllMarkers(merged_seu, only.pos = TRUE,logfc.threshold = 0.1)

fm_fmem <-sum(gc()[, PEAK_MEM_MB])
fm_dmem <- fm_fmem - fm_imem
fm_ed = Sys.time()
fm_time = difftime(fm_ed, fm_st, units = "mins")
cat("FindMarkers took", fm_time,"mins and",
    round(fm_dmem / (1024),5),"GB memory")

```
